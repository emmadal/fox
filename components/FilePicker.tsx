import {
  Modal,
  View,
  Text,
  TouchableOpacity,
  useColorScheme,
  TouchableWithoutFeedback,
} from "react-native";
import { PropsWithChildren } from "react";
import FeatherIcons from "@expo/vector-icons/Feather";
import i18n from "@/i18n";
import { Colors } from "@/constants/Colors";

type Props = PropsWithChildren<{
  isVisible: boolean;
  onClose: () => void;
  onOpenCamera: () => void;
  onTakePhoto: () => void;
}>;

export default function FilePicker({
  isVisible,
  onClose,
  onOpenCamera,
  onTakePhoto,
}: Props) {
  const colorScheme = useColorScheme();

  return (
    <Modal
      animationType="slide"
      transparent={true}
      visible={isVisible}
      onRequestClose={onClose}
      testID="file-picker-modal"
    >
      <TouchableWithoutFeedback onPress={onClose} testID="file-picker-backdrop">
        <View className="flex-1 justify-center items-center bg-transparent">
          <View className="dark:bg-gray-800 bg-gray-100 dark:text-white absolute bottom-1/2 self-center py-6 gap-8 pl-5 w-[90%] rounded-lg">
            <TouchableOpacity
              className="flex-row items-center gap-2"
              testID="camera-button"
              onPress={onOpenCamera}
            >
              <FeatherIcons
                name="camera"
                size={24}
                color={
                  colorScheme === "light" ? Colors.light.text : Colors.dark.text
                }
              />
              <Text className="text-black dark:text-white font-medium text-lg">
                {i18n.t("takephoto")}
              </Text>
            </TouchableOpacity>
            <TouchableOpacity
              className="flex-row items-center gap-2"
              testID="photo-button"
              onPress={onTakePhoto}
            >
              <FeatherIcons
                name="image"
                size={24}
                color={
                  colorScheme === "light" ? Colors.light.text : Colors.dark.text
                }
              />
              <Text className="text-black dark:text-white font-medium text-lg">
                {i18n.t("choosephoto")}
              </Text>
            </TouchableOpacity>
          </View>
        </View>
      </TouchableWithoutFeedback>
    </Modal>
  );
}
