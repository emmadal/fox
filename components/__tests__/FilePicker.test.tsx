import React from "react";
import { render, fireEvent } from "@testing-library/react-native";
import FilePicker from "../FilePicker";

// Mock Feather Icons
jest.mock("@expo/vector-icons/Feather", () => {
  return {
    __esModule: true,
    default: ({ testID }: { testID: string }) => {
      const { View } = require("react-native");
      return <View testID={testID} />;
    },
  };
});

// Mock i18n
jest.mock("@/i18n", () => ({
  t: (key: string) => key,
}));

describe("FilePicker Component", () => {
  const mockProps = {
    isVisible: true,
    onClose: jest.fn(),
    onOpenCamera: jest.fn(),
    onTakePhoto: jest.fn(),
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("renders correctly when visible", () => {
    const { getByTestId, toJSON } = render(<FilePicker {...mockProps} />);

    expect(toJSON()).toMatchSnapshot();
    expect(getByTestId("file-picker-modal")).toBeTruthy();
    expect(getByTestId("file-picker-backdrop")).toBeTruthy();
    expect(getByTestId("camera-button")).toBeTruthy();
    expect(getByTestId("photo-button")).toBeTruthy();
  });

  it("does not render when not visible", () => {
    const { queryByTestId } = render(
      <FilePicker {...mockProps} isVisible={false} />,
    );

    expect(queryByTestId("file-picker-modal")).toBeFalsy();
  });

  it("calls onClose when backdrop is pressed", () => {
    const { getByTestId } = render(<FilePicker {...mockProps} />);

    fireEvent.press(getByTestId("file-picker-backdrop"));
    expect(mockProps.onClose).toHaveBeenCalledTimes(1);
  });

  it("calls onOpenCamera when camera button is pressed", () => {
    const { getByTestId } = render(<FilePicker {...mockProps} />);

    fireEvent.press(getByTestId("camera-button"));
    expect(mockProps.onOpenCamera).toHaveBeenCalledTimes(1);
  });

  it("calls onTakePhoto when photo button is pressed", () => {
    const { getByTestId } = render(<FilePicker {...mockProps} />);

    fireEvent.press(getByTestId("photo-button"));
    expect(mockProps.onTakePhoto).toHaveBeenCalledTimes(1);
  });
});
