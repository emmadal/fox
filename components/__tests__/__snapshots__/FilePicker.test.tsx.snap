// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilePicker Component renders correctly when visible 1`] = `
<Modal
  animationType="slide"
  hardwareAccelerated={false}
  onRequestClose={[MockFunction]}
  testID="file-picker-modal"
  transparent={true}
  visible={true}
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessible={true}
    className="flex-1 justify-center items-center bg-transparent"
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    testID="file-picker-backdrop"
  >
    <View
      className="dark:bg-gray-800 bg-gray-100 dark:text-white absolute bottom-1/2 self-center py-6 gap-8 pl-5 w-[90%] rounded-lg"
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
        testID="camera-button"
      >
        <View />
        <Text
          className="text-black dark:text-white font-medium text-lg"
        >
          takephoto
        </Text>
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
        testID="photo-button"
      >
        <View />
        <Text
          className="text-black dark:text-white font-medium text-lg"
        >
          choosephoto
        </Text>
      </View>
    </View>
  </View>
</Modal>
`;
